token/services/interop/htlc/wallet.go

```diff
@@ -154,6 +241,74 @@ func Wallet(sp view2.ServiceProvider, wallet *token.OwnerWallet, opts ...token.S
 
 	return &OwnerWallet{
 		wallet:       wallet,
+		vault:        vault.TokenVault(),
 		queryService: vault.TokenVault().QueryEngine(),
+		bufferSize:   100,
+	}
+}
+
+type FilteredIterator struct {
+	it   driver.UnspentTokensIterator
+	pick PickFunction
+}
+
+func (f *FilteredIterator) Close() {
+	f.it.Close()
+}
+
+func (f *FilteredIterator) Next() (*token2.UnspentToken, error) {
+	for {
+		tok, err := f.it.Next()
+		if err != nil {
+			return nil, err
+		}
+		if tok == nil {
+			return nil, nil
+		}
+		owner, err := identity.UnmarshallRawOwner(tok.Owner.Raw)
+		logger.Debugf("Is Mine [%s,%s,%s]? No, failed unmarshalling [%s]", view.Identity(tok.Owner.Raw), tok.Type, tok.Quantity, err)
+		if owner.Type == ScriptType {
+			script := &Script{}
+			if err := json.Unmarshal(owner.Identity, script); err != nil {
+				logger.Debugf("token [%s,%s,%s,%s] contains a script? No", tok.Id, view.Identity(tok.Owner.Raw).UniqueID(), tok.Type, tok.Quantity)
+				continue
+			}
+			if script.Sender.IsNone() {
+				logger.Debugf("token [%s,%s,%s,%s] contains a script? No", tok.Id, view.Identity(tok.Owner.Raw).UniqueID(), tok.Type, tok.Quantity)
+				continue
+			}
+			logger.Debugf("token [%s,%s,%s,%s] contains a script? Yes", tok.Id, view.Identity(tok.Owner.Raw).UniqueID(), tok.Type, tok.Quantity)
+
+			pickItem, err := f.pick(tok, script)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to pick (token,script)[%v:%v] pair", tok, script)
+			}
+			if pickItem {
+				return tok, nil
+			}
+		}
+	}
+}
+
+// Sum consumes the iterator and computes the sum of the quantities of the tokens in the iterator
```

Comment:
Sum does not take the iterator as an argument, so I am not sure about the word consumes.

Reply:
Sum is executed over the iterator. This means that after Sum is invoked, if the developer calls again Next, it will get nil.